// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: v1/ranking.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ListProblemRankingsRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListProblemRankingsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListProblemRankingsRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListProblemRankingsRequestMultiError, or nil if none found.
func (m *ListProblemRankingsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListProblemRankingsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Type

	if len(errors) > 0 {
		return ListProblemRankingsRequestMultiError(errors)
	}

	return nil
}

// ListProblemRankingsRequestMultiError is an error wrapping multiple
// validation errors returned by ListProblemRankingsRequest.ValidateAll() if
// the designated constraints aren't met.
type ListProblemRankingsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListProblemRankingsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListProblemRankingsRequestMultiError) AllErrors() []error { return m }

// ListProblemRankingsRequestValidationError is the validation error returned
// by ListProblemRankingsRequest.Validate if the designated constraints aren't met.
type ListProblemRankingsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListProblemRankingsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListProblemRankingsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListProblemRankingsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListProblemRankingsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListProblemRankingsRequestValidationError) ErrorName() string {
	return "ListProblemRankingsRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListProblemRankingsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListProblemRankingsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListProblemRankingsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListProblemRankingsRequestValidationError{}

// Validate checks the field values on ProblemRanking with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ProblemRanking) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ProblemRanking with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ProblemRankingMultiError,
// or nil if none found.
func (m *ProblemRanking) ValidateAll() error {
	return m.validate(true)
}

func (m *ProblemRanking) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Rank

	// no validation rules for UserId

	// no validation rules for Solved

	// no validation rules for Nickname

	if len(errors) > 0 {
		return ProblemRankingMultiError(errors)
	}

	return nil
}

// ProblemRankingMultiError is an error wrapping multiple validation errors
// returned by ProblemRanking.ValidateAll() if the designated constraints
// aren't met.
type ProblemRankingMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ProblemRankingMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ProblemRankingMultiError) AllErrors() []error { return m }

// ProblemRankingValidationError is the validation error returned by
// ProblemRanking.Validate if the designated constraints aren't met.
type ProblemRankingValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ProblemRankingValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ProblemRankingValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ProblemRankingValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ProblemRankingValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ProblemRankingValidationError) ErrorName() string { return "ProblemRankingValidationError" }

// Error satisfies the builtin error interface
func (e ProblemRankingValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sProblemRanking.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ProblemRankingValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ProblemRankingValidationError{}

// Validate checks the field values on ListProblemRankingsResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListProblemRankingsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListProblemRankingsResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListProblemRankingsResponseMultiError, or nil if none found.
func (m *ListProblemRankingsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListProblemRankingsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListProblemRankingsResponseValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListProblemRankingsResponseValidationError{
						field:  fmt.Sprintf("Data[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListProblemRankingsResponseValidationError{
					field:  fmt.Sprintf("Data[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if all {
		switch v := interface{}(m.GetMyRanking()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListProblemRankingsResponseValidationError{
					field:  "MyRanking",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListProblemRankingsResponseValidationError{
					field:  "MyRanking",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetMyRanking()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListProblemRankingsResponseValidationError{
				field:  "MyRanking",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Total

	if len(errors) > 0 {
		return ListProblemRankingsResponseMultiError(errors)
	}

	return nil
}

// ListProblemRankingsResponseMultiError is an error wrapping multiple
// validation errors returned by ListProblemRankingsResponse.ValidateAll() if
// the designated constraints aren't met.
type ListProblemRankingsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListProblemRankingsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListProblemRankingsResponseMultiError) AllErrors() []error { return m }

// ListProblemRankingsResponseValidationError is the validation error returned
// by ListProblemRankingsResponse.Validate if the designated constraints
// aren't met.
type ListProblemRankingsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListProblemRankingsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListProblemRankingsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListProblemRankingsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListProblemRankingsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListProblemRankingsResponseValidationError) ErrorName() string {
	return "ListProblemRankingsResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListProblemRankingsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListProblemRankingsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListProblemRankingsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListProblemRankingsResponseValidationError{}
